src/server/src/main/java/io/cassandrareaper/service/StreamManager.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cassandrareaper.service;
+
+import io.cassandrareaper.AppContext;
+import io.cassandrareaper.ReaperException;
+import io.cassandrareaper.core.Cluster;
+import io.cassandrareaper.core.Node;
+import io.cassandrareaper.core.Stream;
+import io.cassandrareaper.core.StreamSession;
+import io.cassandrareaper.jmx.JmxProxy;
+import io.cassandrareaper.jmx.StreamStatusHandler;
+import io.cassandrareaper.storage.StreamCache;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+import javax.management.openmbean.CompositeData;
+
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.Futures;
+import org.apache.cassandra.streaming.ProgressInfo;
+import org.apache.cassandra.streaming.SessionInfo;
+import org.apache.cassandra.streaming.management.ProgressInfoCompositeData;
+import org.apache.cassandra.streaming.management.SessionInfoCompositeData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class implements the functionality needed for receiving,
+ * handling and serving streaming-related notifications.
+ */
+public final class StreamManager implements StreamStatusHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamManager.class);
+  private final AppContext context;
+  private final ExecutorService executor = Executors.newFixedThreadPool(5);
```

Comment:
I know I've done things this way in the SnapshotManager, but I'm considering the option of making the thread pool size configurable in the yaml file.
It could be necessary to speed things up in large clusters.

What do you think ?

Reply:
I agree with making this configurable. While at it, we can also allow configuring other things, such as size and retention times of the cache holding the stream info.

Reply:
i would think one thread per registered cluster would be enough?

Reply:
this design needs to be concurrent-safe.
but it is not: 
 - pojos should be immutable (progress components separated into separate immutable pojos), 
 - only concurrent collections used (can be enforced by private instantiation)

a number of places could do with guava precondition checks, eg checking against nulls as so forth, eg ensuring concurrency state guarantees.